test/SIM_test_varserv/models/test_client/test_client.cpp

```diff
@@ -0,0 +1,509 @@
+        std::getline(message_stream, token, '\t');
+        std::getline(message_stream, token, ' ');
+        return std::stod(token);
+    };
+
+    // Tail recursion, just for fun
+    std::function<void(int, double)> compare_cycle_forward = [&] (int n_cycles, double last_sim_time) {
+        if (n_cycles <= 0)
+            return;
+
+        double sim_time = parse_message_for_sim_time(socket.receive());
+        EXPECT_LE(abs((sim_time - last_sim_time) - cycle), tolerance);
+        compare_cycle_forward(n_cycles-1, sim_time);
+    };
+
+    auto run_cycle_test = [&] () {
+        std::string command = "trick.var_cycle(" + std::to_string(cycle) + ")\n";
+        socket << command;
+        double sim_time = parse_message_for_sim_time(socket.receive());
+        compare_cycle_forward(num_cycles, sim_time);
+    };
+
+    std::string command = "trick.var_add(\"time\")\n";
+    socket << command;
+
+    tolerance = 0.011;
+    cycle = 1.0;
+    run_cycle_test();
+    
+    cycle = 0.5;
+    run_cycle_test();
+
+    tolerance = 0.015;
+    cycle = 0.1;    
+    run_cycle_test();
+
+    num_cycles = 2;
+    cycle = 3.0;
+    tolerance = 0.1;
+    run_cycle_test();
+}
+
+
+TEST_F (VariableServerTest, Pause) {
+    if (socket_status != 0) {
+        FAIL();
+    }
+    
+    std::string reply;
+    std::string expected;
+
+    socket << "trick.var_add(\"vsx.vst.f\")\ntrick.var_add(\"vsx.vst.i\")\ntrick.var_pause()\n";
+
+    // Make sure it's paused
+    EXPECT_EQ(socket.check_for_message_availible(), false);
+
+    socket << "trick.var_send()\n";
+    socket >> reply;
+    expected = std::string("0  123456  1234.5677");
+
+    EXPECT_EQ(strcmp_IgnoringWhiteSpace(reply, expected), 0);
+
+    // Make sure it doesn't send cyclically
+    EXPECT_EQ(socket.check_for_message_availible(), false);
+
+    socket << "trick.var_unpause()\n";
+    socket >> reply;
+    expected = std::string("0  123456  1234.5677");
+
+    EXPECT_EQ(strcmp_IgnoringWhiteSpace(reply, expected), 0);
+}
+
+TEST_F (VariableServerTest, CopyAndWriteModes) {
+    if (socket_status != 0) {
+        FAIL();
+    }
+
+    std::string reply;
+    std::string expected;
+
+    // We're just checking that every combination of modes is functional
+    // We can't test that they perform their copying and writing in the correct place from here
+    // Default is 0 0
+    socket << "trick.var_add(\"vsx.vst.a\")\ntrick.var_add(\"vsx.vst.b\")\n";
+    socket >> reply;
+
+    expected = "0 97 98";
+    EXPECT_EQ(strcmp_IgnoringWhiteSpace(reply, expected), 0);
+
+    // Clear out anything else that's been sent
+    // I may need to write something else for this
+    socket << "trick.var_pause()\n";
+    socket.clear_buffered_data();
+
+    // Copy mode 1 (VS_COPY_SCHEDULED) write mode 0 (VS_WRITE_ASYNC)
+    socket << "trick.var_clear()\ntrick.var_set_copy_mode(1)\ntrick.var_add(\"vsx.vst.c\")\ntrick.var_add(\"vsx.vst.d\")\ntrick.var_unpause()\n";
+    socket >> reply;
+
```

Comment:
The behavior I'm seeing in this test is that when you use copy mode VS_COPY_SCHEDULED or VS_COPY_TOP_OF_FRAME with write mode VS_WRITE_ASYNC, the first message may have all 0 for data since the first copy hadn't occurred before the first write, and there's no synchronization preventing unstaged data from going out. Is this what we want? Should I preserve this behavior in the refactor or fix it?